### PR TITLE
Dask array slicing with boolean arrays

### PR DESCRIPTION
**Pull request by mrocklin**
Currently we support this:

    x[y > 0]

But not this

    x[:, y> 0]

This PR attempts to fix this, but currently fails.  

see #2651 and #2430 

cc @TomAugspurger

### CODE REVIEW

**Comment by jakirkham** [comment]
As noted in the associated issue, it may be worth running `reshape` to flatten the dimensions that will be selected from and pulling code from `compress` to apply this selection mask on the flattened dimensions.

**Comment by mrocklin** [comment]
I've added an implementation.  However I had forgotten that NumPy does some odd things when slicing with boolean arrays.  In particular it seems to prefer to bring that axis to the front.

**Comment by mrocklin** [comment]
Not sure what to do here.

**Comment by jakirkham** [comment]
> However I had forgotten that NumPy does some odd things when slicing with boolean arrays. In particular it seems to prefer to bring that axis to the front.

Sorry not following. Do you have an example?

**Comment by mrocklin** [comment]
I'm finding this behavior to be confusing:

```python
import numpy as np
x = np.ones((10, 10, 10, 10))
x[1:9, 4, :, np.array([True, False, False, True, False, False, True, True, True, False])].shape

# (5, 8, 10)
```

This PR has an xfailed test that includes this situation

**Comment by shoyer** [comment]
This is numpy's confusing mixed basic/advanced indexing rules again. You get the same sort of behavior with integer instead of boolean arrays.

**Comment by shoyer** [comment]
One way to resolve this is to add a dedicated `oindex` method for outer/orthogonal indexing, as [proposed for numpy](https://github.com/numpy/numpy/pull/6256). Then the confusing case (with "advanced" indexes separated by basic indexes) could be explicitly disallowed. See https://docs.scipy.org/doc/numpy-1.13.0/reference/arrays.indexing.html#combining-advanced-and-basic-indexing for the full details

**Comment by jakirkham** [comment]
Have to say am a bit perplexed by that example as well. Would have to agree that it looks like it is transposing things when it shouldn't. Are we sure it isn't a bug?

**Comment by shoyer** [comment]
Definitely not a bug, but you're not the only one who's confused by it. (There have been long discussions about the numpy mailing list about this.)

**Comment by mrocklin** [comment]
This is ready for review

File: dask/array/slicing.py [hunk 1]
```diff
@@ -813,3 +826,57 @@ def check_index(ind, dimension):
...
+        y = elemwise(getitem, x, *index, dtype=x.dtype)
+        name = 'getitem-' + tokenize(x, index)
+        dsk = {(name, i): k for i, k in enumerate(core.flatten(y._keys()))}
+        chunks = ((np.nan,) * y.npartitions,)
+        return (Array(sharedict.merge(y.dask, (name, dsk)), name, chunks, x.dtype),
+                out_index)
+
+    if any(isinstance(ind, Array) and ind.dtype == bool and ind.ndim != 1
+            for ind in index):
+        raise NotImplementedError("Slicing with dask.array only permitted for "
```
Comment (TomAugspurger) [inline dask/array/slicing.py]: I found this error message a bit confusing, I think the "only permitted for dimension 1 and dimension n". Maybe something like "only permitted when the indexer has 1 dimension, or the dimensions of the indexer match the dimensions of the array being sliced."? It's a bit wordier.

Comment (mrocklin) [reply dask/array/slicing.py]: Fixed

**Pull request merged**